File: documentation/examples/ping_pong/pong.cc
#include "absl/flags/flag.h"
#include "absl/log/check.h"
#include "absl/log/log.h"

#include "aos/configuration.h"
#include "aos/events/shm_event_loop.h"
#include "aos/init.h"
#include "documentation/examples/ping_pong/ping_generated.h"
#include "documentation/examples/ping_pong/pong_generated.h"
#include "documentation/examples/ping_pong/pong_lib.h"

ABSL_FLAG(std::string, config, "pingpong_config.json", "Path to the config.");

int main(int argc, char **argv) {
  aos::InitGoogle(&argc, &argv);
  aos::EventLoop::SetDefaultVersionString("pong_version");

  aos::FlatbufferDetachedBuffer<aos::Configuration> config =
      aos::configuration::ReadConfig(absl::GetFlag(FLAGS_config));

  ::aos::ShmEventLoop event_loop(&config.message());

  aos::Pong ping(&event_loop);

  event_loop.Run();

  return 0;
}

File: documentation/examples/ping_pong/pong.py
import flatbuffers
from absl import app
from absl import logging

import aos.events.util as util
import ping_fbs_py.aos.examples.Ping as PingFbs
import pong_fbs_py.aos.examples.Pong as PongFbs
from aos.events.shm_event_loop import ShmEventLoop


async def pong(runtime):
    fetcher = runtime.make_fetcher("/test", PingFbs.Ping)
    sender = runtime.make_sender("/test", PongFbs.Pong)
    await runtime.on_run()
    while True:
        ping = await fetcher.next()
        fetched, fetched_ping = fetcher.fetch()
        assert fetched
        assert ping.Value() == fetched_ping.Value()
        logging.vlog(1, f"Received Ping({ping.Value()})")
        fbb = flatbuffers.Builder()
        PongFbs.Start(fbb)
        PongFbs.AddValue(fbb, ping.Value())
        PongFbs.AddInitialSendTime(fbb, ping.SendTime())
        fbb.Finish(PongFbs.End(fbb))
        sender.send(fbb.Output())


def run_pong(runtime):
    # TODO(Sanjay): Uncomment the following line once the C API is implemented for it.
    # runtime.set_realtime_priority(5)
    runtime.spawn_task(pong(runtime))


def main(argv):
    util.init(argv)
    config = util.Configuration("aos/aos/events/pingpong_config.bfbs")
    with ShmEventLoop(config) as shm_event_loop:
        shm_event_loop.run_with(run_pong)


if __name__ == "__main__":
    app.run(main)
